packages/expo-updates/ios/EXUpdates/EXUpdatesDatabase.m

```diff
@@ -0,0 +1,376 @@
+    return nil;
+  }
+  if (args) {
+    [self _bindStatement:stmt withArgs:args];
+  }
+
+  NSMutableArray *rows = [NSMutableArray arrayWithCapacity:0];
+  NSMutableArray *columnNames = [NSMutableArray arrayWithCapacity:0];
+  NSString *errorMessage;
+
+  int columnCount = 0;
+  BOOL fetchedColumns = NO;
+  int result;
+  BOOL hasMore = YES;
+  while (hasMore) {
+    result = sqlite3_step (stmt);
+    switch (result) {
+      case SQLITE_ROW: {
+        if (!fetchedColumns) {
+          // get all column names once at the beginning
+          columnCount = sqlite3_column_count(stmt);
+          
+          for (int i = 0; i < columnCount; i++) {
+            [columnNames addObject:[NSString stringWithFormat:@"%s", sqlite3_column_name(stmt, i)]];
+          }
+          fetchedColumns = YES;
+        }
+        NSMutableDictionary *entry = [NSMutableDictionary dictionary];
+        for (int i = 0; i < columnCount; i++) {
+          NSObject *columnValue = [self _getValueWithStatement:stmt column:i];
+          entry[columnNames[i]] = columnValue;
+        }
+        [rows addObject:entry];
+        break;
+      }
+      case SQLITE_DONE:
+        hasMore = NO;
+        break;
+      default:
+        errorMessage = [[self class] convertSQLiteErrorToString:_db];
+        hasMore = NO;
+        break;
+    }
+  }
+
+  sqlite3_finalize(stmt);
+  
+  if (errorMessage) {
+    [[EXUpdatesAppController sharedInstance] handleErrorWithDomain:kEXUpdatesDatabaseErrorDomain description:errorMessage info:nil isFatal:NO];
+    return nil;
+  }
+  
+  return rows;
+}
+
+- (id)_getValueWithStatement:(sqlite3_stmt *)stmt column:(int)column
+{
+  int columnType = sqlite3_column_type(stmt, column);
+  switch (columnType) {
+    case SQLITE_INTEGER:
+      return @(sqlite3_column_int64(stmt, column));
+    case SQLITE_FLOAT:
+      return @(sqlite3_column_double(stmt, column));
+    case SQLITE_BLOB:
+      return [[NSUUID alloc] initWithUUIDBytes:sqlite3_column_blob(stmt, column)];
+    case SQLITE_TEXT:
+      return [[NSString alloc] initWithBytes:(char *)sqlite3_column_text(stmt, column)
+                                      length:sqlite3_column_bytes(stmt, column)
+                                    encoding:NSUTF8StringEncoding];
+  }
+  return [NSNull null];
+}
+
+- (void)_bindStatement:(sqlite3_stmt *)stmt withArgs:(NSArray *)args
+{
+  [args enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
+    if ([obj isKindOfClass:[NSUUID class]]) {
+      uuid_t bytes;
+      [((NSUUID *)obj) getUUIDBytes:bytes];
+      if (sqlite3_bind_blob(stmt, (int)idx + 1, bytes, 16, SQLITE_TRANSIENT) != SQLITE_OK) {
+        [[EXUpdatesAppController sharedInstance] handleErrorWithDomain:kEXUpdatesDatabaseErrorDomain description:@"Could not bind UUID to SQLite statement" info:nil isFatal:NO];
```

Comment:
Generally speaking, I think we want to avoid global side effects (acting on the `sharedInstance`) and as a principle, "flatter is better than nested" -- in this case, lifting this side effect of error reporting up somewhere higher in the call stack and returning (probably via an "out" pointer) an error instead.